types/ramda/test/when-tests.ts

```diff
@@ -1,25 +1,33 @@
 import * as R from 'ramda';
 
 () => {
-  // $ExpectType (a: string) => string
-  const truncate = R.when(
-    (str: string) => str.length > 10,
-    (str: string) => str.slice(0, 10) + '…',
-  );
+    // $ExpectType <A extends string = string>(a: A) => string | A
+    const truncate = R.when(
+        (str: string) => str.length > 10,
+        (str: string) => str.slice(0, 10) + '…',
+    );
 
-  // $ExpectType string
-  truncate('12345'); // => '12345'
-  // $ExpectType string
-  truncate('0123456789ABC'); // => '0123456789…'
+    // $ExpectType string
+    truncate('12345'); // => '12345'
 
-  // $ExpectType (a: number | undefined) => number | undefined
-  const addOneIfNotNil = R.when<undefined | number, number>(
-      x => x != null,
-      R.add(1)
-  );
+    // $ExpectType string
+    truncate('0123456789ABC'); // => '0123456789…'
 
-  // $ExpectType number | undefined
-  const nil = addOneIfNotNil(undefined);
-  // $ExpectType number | undefined
-  const two = addOneIfNotNil(1);
+    // $ExpectType <A extends number | undefined = number>(a: A) => string | A
+    const testBroaderType = R.when(
+        (x: number | undefined) => x != null,
+        (x: number | string) => 'some text',
+    );
+
+    // $ExpectType <A extends any = any>(a: A) => number | A
+    const addOneIfNotNil = R.when(
+        R.isNil,
+        R.add(1)
+    );
+
+    // $ExpectType number | undefined
+    const nil = addOneIfNotNil(undefined);
```

Comment:
```typescript
const nil = addOneIfNotNil("1");
```
will give you a return type of `number | "1"`, which is both wrong and shouldn't be valid anyway

Reply:
Since isNil accepts `any` and `any & number => any` so return functions allow anything, but that is because of typing of `isNull`. With old version there is no sense using it if the type will be always `number` then isNill will be always `false`, so this usage is no particular impact anyway.
The benefits are:
1: when you wanna pass typed function, expressin `P & Parameters<T>[0]` is gonna take intersection(and that is logically) so test `testBroaderType` will work, but with current version it does not work:
```ts
const testBroaderType = R.when(
        (x: number | undefined) => x != null,
        (x: number | string) => 'some text',
    );
```
above we will have (number) => string | number

and if you wanna reuse some functions where one does not extend another(as in the case above) it is not gonna work in current `when`, but such cases are possible to use in real app.

2: you can do something like const `nil = addOneIfNotNil<string>("1");` in general sense or even:
```ts
const test2= R.when(
        (x: number | undefined) => x != null,
        (x: number | string) => 'some text',
    )<number>;
```
3: works with `strict` typescript compiler option